File: problem_2.cpp
#include <chrono>
#include <iostream>
#include <vector>

const long long limitNumber = 4000000;

class Fibonacci 
{
    std::vector<long> fibCache;
    const int cachePresizing = 100;

    public:
    Fibonacci() 
    {
        fibCache.resize(cachePresizing);
	for (int i = 0; i < cachePresizing; i++) 
	{
            fibCache[i] = 0; 
	}
    }

    ~Fibonacci() 
    {
        fibCache.resize(0);
    }

    long fib(int n)
    {
        if (n <= 1) 
        {
            fibCache[n] = long(1);
            return long(1);
        }
	if (fibCache[n] != 0) 
	{
            return fibCache[n];
	}
	long prev2 = (fibCache[n - 2] != 0) ? fibCache[n - 2] : fib(n - 2);
	long prev1 = (fibCache[n - 1] != 0) ? fibCache[n - 1] : fib(n - 1);
	fibCache[n] = prev2 + prev1;
	return prev2 + prev1;
    }
};


int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();
    
    Fibonacci computeFib;
    long long runningSum = 0;
    int i = 0;
    int curFib = 0;
    do {
        curFib = computeFib.fib(i);
	if (curFib % 2 == 0) 
	{
	    runningSum += curFib;
	}
	i++;
    }
    while (curFib < limitNumber);

    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << runningSum << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_2.rs
use std::time::{Instant};

const LIMIT_NUMBER: usize = 4000000;
const CACHE_PRESIZE: usize = 100;

struct Fibonacci 
{
    fib_cache : Vec<usize>,
}

impl Fibonacci 
{
    fn init () -> Fibonacci 
    {
        let mut ret = Fibonacci { fib_cache: vec![] };
        ret.fib_cache.resize(CACHE_PRESIZE, 0);
        ret
    }

    fn clean_cache (&mut self) 
    {
        self.fib_cache.resize(0, 0);
    }

    fn fib (&mut self, n : usize) -> usize 
    {
        if n <= 1
        {
            self.fib_cache[n] = 1;
            return 1;
        }
        if self.fib_cache[n] != 0 { return self.fib_cache[n] };
        let prev2 = if self.fib_cache[n - 2] != 0 { self.fib_cache[n - 2] } else { self.fib(n - 2) };
        let prev1 = if self.fib_cache[n - 1] != 0 { self.fib_cache[n - 1] } else { self.fib(n - 1) };
        self.fib_cache[n] = prev2 + prev1;
        return prev2 + prev1;
    }
}

fn main () 
{
    let start = Instant::now();

    let mut fib_with_cache = Fibonacci::init();
    let mut final_sum = 0;
    let mut cur_fib = 0;
    let mut i = 1;
    while cur_fib < LIMIT_NUMBER 
    {
        cur_fib = fib_with_cache.fib(i);
        if cur_fib % 2 == 0
        {
            final_sum += cur_fib;
        }
        i = i + 1;
    }
    fib_with_cache.clean_cache();

    let duration = start.elapsed();
    println!("Result: {}", final_sum);
    println!("Timing: {}", duration.as_nanos());
}